File: sahayog/sahayog/report/approval_request_analysis/approval_request_analysis.js
frappe.query_reports["Approval Request Analysis"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname": "employee",
			"label": __("Employee"),
			"fieldtype": "Link",
			"options": "Employee"
		},
		{
			"fieldname": "category",
			"label": __("Category"),
			"fieldtype": "Link",
			"options": "Approval Category"
		},
		{
			"fieldname": "approval_status",
			"label": __("Status"),
			"fieldtype": "Select",
			"options": "\nDraft\nPending Approval\nApproved\nRejected"
		}
	],

	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "approval_status") {
			const status_colors = {
				"Approved": "green",
				"Pending Approval": "orange",
				"Rejected": "red",
				"Draft": "gray"
			};
			const color = status_colors[data.approval_status] || "gray";
			value = `<span class="indicator-pill ${color}">${data.approval_status}</span>`;
		}

		return value;
	},

	"onload": function(report) {
		// Force styles after a small delay to ensure DataTable has rendered
		setTimeout(() => {
			frappe.dom.set_style(`
				/* Force Horizontal Scroll and prevent squeezing */
				.frappe-list .result-list, .dt-instance-1 {
					overflow-x: auto !important;
				}

				.data-table {
					table-layout: auto !important;
					width: max-content !important;
					min-width: 100% !important;
				}

				/* Target cells directly to force widths */
				.dt-cell {
					white-space: nowrap !important;
					display: flex !important;
					align-items: center !important;
					border-right: 1px solid #f0f0f0 !important;
					min-width: 150px !important; /* Default min-width */
				}

				/* Specific Widths based on column index (as fieldname targeting is tricky in DT) */
				.dt-cell[data-col-index="4"] { min-width: 400px !important; } /* Title */
				.dt-cell[data-col-index="8"] { min-width: 500px !important; } /* Remark */

				.dt-header .dt-cell {
					font-weight: bold !important;
					background-color: #f8f9fa !important;
				}
				
				/* Fix header/body alignment sync */
				.dt-header, .dt-body {
					width: max-content !important;
					min-width: 100% !important;
				}
			`);
		}, 500);
	}
};

File: sahayog/sahayog/report/approval_request_analysis/approval_request_analysis.py
import frappe
from frappe import _
from sahayog.sahayog.doctype.approval_request.approval_request import get_permission_query_conditions

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": _("Request ID"), "fieldname": "name", "fieldtype": "Link", "options": "Approval Request", "width": 300},
        {"label": _("Employee"), "fieldname": "employee_name", "fieldtype": "Data", "width": 300},
        {"label": _("Designation"), "fieldname": "designation", "fieldtype": "Data", "width": 300},
        {"label": _("Category"), "fieldname": "category", "fieldtype": "Data", "width": 300},
        {"label": _("Title"), "fieldname": "title", "fieldtype": "Data", "width": 400},
        {"label": _("Status"), "fieldname": "approval_status", "fieldtype": "Data", "width": 300},
        {"label": _("Acted By"), "fieldname": "acted_by", "fieldtype": "Data", "width": 300},
        {"label": _("Creation Date"), "fieldname": "creation", "fieldtype": "Datetime", "width": 300},
        {"label": _("Approver Remark"), "fieldname": "approver_remark", "fieldtype": "Small Text", "width": 500}
    ]

def get_data(filters):
    conditions = get_conditions(filters)
    
    # Permission Condition
    perm_cond = get_permission_query_conditions(frappe.session.user)
    if perm_cond:
        conditions.append(f"({perm_cond})")
    
    where_clause = " AND ".join(conditions) if conditions else "1=1"
    
    # Query fields order MUST match columns order
    query = f"""
        SELECT 
            name, employee_name, designation, category, title, 
            approval_status, acted_by, creation, approver_remark
        FROM 
            `tabApproval Request`
        WHERE 
            {where_clause}
        ORDER BY 
            creation DESC
    """
    
    return frappe.db.sql(query, filters, as_dict=True)

def get_conditions(filters):
    conditions = []
    
    if filters.get("from_date"):
        conditions.append("creation >= %(from_date)s")
    if filters.get("to_date"):
        filters["to_date_end"] = f"{filters.get('to_date')} 23:59:59"
        conditions.append("creation <= %(to_date_end)s")
    if filters.get("employee"):
        conditions.append("employee = %(employee)s")
    if filters.get("category"):
        conditions.append("category = %(category)s")
    if filters.get("approval_status"):
        conditions.append("approval_status = %(approval_status)s")
        
    return conditions
